src/online2bin/online2-net-nnet3-latgen-faster.cc

```diff
@@ -0,0 +1,398 @@
+                  << word_syms_rxfilename;
+
+    KALDI_LOG << "Loaded evertyhing. Waiting for data...";
+
+    TcpServer server;
+
+    server.Listen(port_num);
+
+    while (true) {
+
+      server.Accept();
+
+      int32 samp_count = 0;
+      size_t chunk_len = static_cast<size_t>(chunk_length_secs * samp_freq);
+
+      int32 check_update = static_cast<int32>(samp_freq * output_freq);
+      int32 next_check = check_update;
+
+      bool eos = false;
+
+      OnlineIvectorExtractorAdaptationState adaptation_state(
+          feature_info.ivector_extractor_info);
+
+      while (!eos) {
+
+        OnlineNnet2FeaturePipeline feature_pipeline(feature_info);
+        feature_pipeline.SetAdaptationState(adaptation_state);
+
+        OnlineSilenceWeighting silence_weighting(
+            trans_model,
+            feature_info.silence_weighting_config,
+            decodable_opts.frame_subsampling_factor);
+
+        SingleUtteranceNnet3Decoder decoder(decoder_opts, trans_model,
+                                            decodable_info,
+                                            *decode_fst, &feature_pipeline);
+
+        std::vector<std::pair<int32, BaseFloat> > delta_weights;
+
+        while (true) {
+
+          eos = !server.ReadChunk(chunk_len);
+
+          if (!eos) {
+            Vector<BaseFloat> wave_part = server.getChunk();
+            feature_pipeline.AcceptWaveform(samp_freq, wave_part);
+            samp_count += chunk_len;
+
+            if (silence_weighting.Active() &&
+                feature_pipeline.IvectorFeature() != NULL) {
+              silence_weighting.ComputeCurrentTraceback(decoder.Decoder());
+              silence_weighting.GetDeltaWeights(feature_pipeline.NumFramesReady(),
+                                                &delta_weights);
+              feature_pipeline.IvectorFeature()->UpdateFrameWeights(delta_weights);
+            }
+
+            decoder.AdvanceDecoding();
+
+            if (samp_count > next_check) {
+              if (decoder.NumFramesDecoded() > 0) {
+                Lattice lat;
+                decoder.GetBestPath(false, &lat);
+
+                std::string msg = LatticeToString(lat, word_syms);
+
+                server.Write(msg + '\r');
+              }
+
+              next_check += check_update;
+            }
+
+          } else {
+            feature_pipeline.InputFinished();
+
+            decoder.FinalizeDecoding();
+
+            if (decoder.NumFramesDecoded() > 0) {
+              Lattice lat;
+              decoder.GetBestPath(true, &lat);
+
+              std::string msg = LatticeToString(lat, word_syms);
+
+              server.Write(msg + "\n");
+            }
+
+            server.disconnect();
+            break;
+
+          }
+
+          if (decoder.EndpointDetected(endpoint_opts)) {
+            server.Write("\n");
+            feature_pipeline.GetAdaptationState(&adaptation_state);
+            break;
```

Comment:
Are you sure no part of the output is missed? When breaking here, it seems to me that there might be new results that have not been sent to the client yet (e.g. if `check_update` is large).

Reply:
...and BTW, the decoder is going to be destroyed here without `FinalizeDecoding` having been called on it.

Reply:
I'll have to think about this a bit more. I'll get back to you on that ASAP.

Reply:
Ok, I made sure to finalize decoding after every speech segment and output the real best path from the final lattice.

I do realize a small semantic problem here. I used the name from the other binary (online2-wav-nnet3-latgen-faster), but I'm not really interested in generating a lattice here (and sending it over TCP). This program is intended more to preview the decoder output in real-time. Do you think I should rename it to something like "online2-net-nnet3-decode-faster"?

Reply:
That makes sense. Actually, it might be better to include TCP in the name-- something like `online2-nnet3-tcp-decoder` or `-server`.

Reply:
Ok, I renamed it to "online2-tcp-nnet3-decode-faster" so it syncs up with other names better.